File: src/Interp.Python/zeta_interp/relational_identity.py
"""Independent exact regeneration of the registered relational-receipt panel.

The verifier is a fixture table, never a cryptographic implementation. This
module reads no native outputs while regenerating its expected semantic panel.
"""

from __future__ import annotations

import argparse
import json
from collections import Counter, defaultdict
from dataclasses import asdict, dataclass, replace
from fractions import Fraction
from hashlib import sha256
from itertools import combinations, permutations, product
from pathlib import Path
from typing import Any


@dataclass(frozen=True, order=True)
class Receipt:
    EventId: str
    Actor: str
    Counterparty: str
    Interaction: str
    Channel: str
    Claim: str
    Weight: int
    Parents: tuple[str, ...]


@dataclass(frozen=True)
class Occurrence:
    position: int
    clock: int
    receipt: Receipt
    attestation: str


@dataclass(frozen=True)
class View:
    observer: str
    occurrences: tuple[Occurrence, ...]


RECEIPTS = (
    Receipt("e0", "A", "B", "i0", "vision", "color", 1, ()),
    Receipt("e1", "A", "B", "i1", "vision", "color", -1, ("e0",)),
    Receipt("e2", "B", "A", "i0", "vision", "color", 1, ("e0",)),
    Receipt("e3", "A", "B", "i2", "receipt", "ack", 1, ("e1", "e2")),
)
EXPECTED = tuple(r.EventId for r in RECEIPTS)
TABLE = {(r, "fixture-" + r.EventId) for r in RECEIPTS}


def view(observer: str, receipts: tuple[Receipt, ...]) -> View:
    return View(
        observer,
        tuple(
            Occurrence(i, i, r, "fixture-" + r.EventId) for i, r in enumerate(receipts)
        ),
    )


def authenticate(table: set[tuple[Receipt, str]], item: Occurrence) -> str:
    if item.attestation == "unavailable":
        return "unavailable"
    return "accepted" if (item.receipt, item.attestation) in table else "rejected"


def wire_invariant(
    receipts: set[Receipt], pairs: set[tuple[str, str]]
) -> dict[str, Any]:
    balances: Counter[tuple[str, str, str, str]] = Counter()
    for r in receipts:
        balances[r.Actor, r.Counterparty, r.Channel, r.Claim] += r.Weight
    return {
        "Receipts": [
            {**asdict(r), "Parents": list(r.Parents)} for r in sorted(receipts)
        ],
        "CausalPairs": [list(p) for p in sorted(pairs)],
        "Claims": [
            list(key) + [str(weight)]
            for key, weight in sorted(balances.items())
            if weight
        ],
    }


def compare(
    name: str,
    table: set[tuple[Receipt, str]],
    cut: tuple[str, ...],
    left: View,
    right: View,
) -> dict[str, Any]:
    out: dict[str, Any] = {
        "Name": name,
        "Status": "",
        "RepeatedLeft": 0,
        "RepeatedRight": 0,
        "Invariant": [],
    }
    for field in (
        "MissingLeft",
        "MissingRight",
        "UnverifiedLeft",
        "UnverifiedRight",
        "RejectedLeft",
        "RejectedRight",
        "Conflicts",
        "BoundaryParents",
    ):
        out[field] = []
    cut_set = set(cut)
    for v in (left, right):
        if len(v.occurrences) > 256 or any(
            o.receipt.EventId not in cut_set for o in v.occurrences
        ):
            out["Status"] = "refused-input"
            return out
        if len({o.position for o in v.occurrences}) != len(v.occurrences) or any(
            o.position < 0 for o in v.occurrences
        ):
            out["Status"] = "refused-coordinates"
            return out
    accepted_views: list[list[Occurrence]] = []
    union: set[Receipt] = set()
    unknown = False
    rejected = False
    for suffix, v in (("Left", left), ("Right", right)):
        normalized = [
            replace(
                o,
                receipt=replace(
                    o.receipt, Parents=tuple(sorted(set(o.receipt.Parents)))
                ),
            )
            for o in v.occurrences
        ]
        accepted = [o for o in normalized if authenticate(table, o) == "accepted"]
        accepted_views.append(accepted)
        accepted_ids = {o.receipt.EventId for o in accepted}
        unverified = {
            o.receipt.EventId
            for o in normalized
            if authenticate(table, o) == "unavailable"
        } - accepted_ids
        invalid = {
            o.receipt.EventId
            for o in normalized
            if authenticate(table, o) == "rejected"
        }
        out["Missing" + suffix] = sorted(
            cut_set - {o.receipt.EventId for o in normalized}
        )
        out["Unverified" + suffix] = sorted(unverified)
        out["Rejected" + suffix] = sorted(invalid)
        out["Repeated" + suffix] = len(accepted) - len({o.receipt for o in accepted})
        unknown |= not cut_set <= accepted_ids
        rejected |= bool(invalid)
        union.update(o.receipt for o in accepted)
    grouped: dict[str, set[Receipt]] = defaultdict(set)
    parents: dict[str, set[str]] = defaultdict(set)
    for r in union:
        grouped[r.EventId].add(r)
        parents[r.EventId].update(r.Parents)
    out["Conflicts"] = sorted(
        event for event, contents in grouped.items() if len(contents) > 1
    )
    out["BoundaryParents"] = sorted(
        {p for values in parents.values() for p in values} - cut_set
    )
    unknown |= bool(out["BoundaryParents"])
    # DFS from each event provides an algorithm independent of the native pivot closure.
    pairs: set[tuple[str, str]] = set()
    for target in grouped:
        frontier = list(parents[target])
        visited: set[str] = set()
        while frontier:
            ancestor = frontier.pop()
            if ancestor in visited or ancestor not in grouped:
                continue
            visited.add(ancestor)
            pairs.add((ancestor, target))
            frontier.extend(parents[ancestor])
    cycle = any(a == b for a, b in pairs)
    if not out["Conflicts"] and not cycle:
        for items in accepted_views:
            positions: dict[str, int] = {}
            for o in items:
                positions[o.receipt.EventId] = min(
                    o.position, positions.get(o.receipt.EventId, o.position)
                )
            if any(
                a in positions and b in positions and positions[a] >= positions[b]
                for a, b in pairs
            ):
                empty = compare(name, set(), (), View("A", ()), View("B", ()))
                empty.update(Status="refused-coordinates", Invariant=[])
                return empty
    if out["Conflicts"]:
        out["Status"] = "authenticated-conflict"
    elif cycle:
        out["Status"] = "authenticated-causal-cycle"
    elif rejected:
        out["Status"] = "authentication-rejected"
    elif unknown:
        out["Status"] = "unknown-coverage"
    else:
        out["Status"] = "consistent-on-declared-cut"
        out["Invariant"] = [wire_invariant(union, pairs)]
    return out


def transport_panel() -> dict[str, Any]:
    left = view("A", RECEIPTS)
    invariant = compare("", TABLE, EXPECTED, left, left)["Invariant"][0]
    arrivals = list(permutations(EXPECTED))
    topologies = [
        order
        for order in arrivals
        if all(
            order.index(parent) < order.index(r.EventId)
            for r in RECEIPTS
            for parent in r.Parents
        )
    ]
    by_id = {r.EventId: r for r in RECEIPTS}
    checks = 0
    for order in topologies:
        original = view("B", tuple(by_id[event] for event in order))
        by_event = {o.receipt.EventId: o for o in original.occurrences}
        for arrival, offset, scale in product(arrivals, (-1000, 0, 1000), (1, 7)):
            shuffled = View("B", tuple(by_event[event] for event in arrival))
            moved = View(
                "B",
                tuple(
                    replace(
                        o, position=3 * o.position + 5, clock=offset + scale * o.clock
                    )
                    for o in shuffled.occurrences
                ),
            )
            assert compare("", TABLE, EXPECTED, left, moved)["Invariant"] == [invariant]
            restored = View(
                "B",
                tuple(
                    replace(
                        o,
                        position=(o.position - 5) // 3,
                        clock=(o.clock - offset) // scale,
                    )
                    for o in moved.occurrences
                ),
            )
            assert restored == shuffled
            intermediate = tuple(
                replace(o, position=o.position + 3, clock=o.clock + 11)
                for o in moved.occurrences
            )
            successive = View(
                "B",
                tuple(
                    replace(o, position=7 * o.position, clock=5 * o.clock)
                    for o in intermediate
                ),
            )
            direct = View(
                "B",
                tuple(
                    replace(o, position=7 * (o.position + 3), clock=5 * (o.clock + 11))
                    for o in moved.occurrences
                ),
            )
            assert successive == direct
            assert compare("", TABLE, EXPECTED, left, direct)["Invariant"] == [
                invariant
            ]
            checks += 1
    return {
        "ArrivalOrders": len(arrivals),
        "TopologicalOrders": [list(o) for o in topologies],
        "Checks": checks,
        "InverseChecks": checks,
        "CompositionChecks": checks,
        "Invariant": invariant,
    }


def mutation_panel() -> list[dict[str, Any]]:
    left, right = view("A", RECEIPTS), view("B", RECEIPTS)

    def update(v: View, event: str, **fields: Any) -> View:
        return replace(
            v,
            occurrences=tuple(
                replace(o, **fields) if o.receipt.EventId == event else o
                for o in v.occurrences
            ),
        )

    def omit(v: View, event: str) -> View:
        return replace(
            v, occurrences=tuple(o for o in v.occurrences if o.receipt.EventId != event)
        )

    changed = replace(RECEIPTS[2], Claim="green")
    forked = update(right, "e2", receipt=changed, attestation="fixture-fork")
    fork_table = TABLE | {(changed, "fixture-fork")}
    cyclic = replace(RECEIPTS[0], Parents=("e3",))
    cycle_table = TABLE | {(cyclic, "fixture-cycle")}
    boundary = replace(RECEIPTS[0], Parents=("outside",))
    boundary_table = TABLE | {(boundary, "fixture-boundary")}
    alternate = replace(RECEIPTS[2], EventId="e4", Channel="auditory")
    extended = RECEIPTS + (alternate,)
    fabricated = tuple(
        Receipt(
            f"f{actor}-{counterparty}",
            f"fake{actor}",
            f"fake{counterparty}",
            f"pair{a}-{b}",
            "fixture",
            "exchange",
            1,
            (),
        )
        for a, b in combinations(range(4), 2)
        for actor, counterparty in ((a, b), (b, a))
    )
    fabricated_table = {(r, "fixture-" + r.EventId) for r in fabricated}
    return [
        compare("baseline", TABLE, EXPECTED, left, right),
        compare(
            "identical-replay",
            TABLE,
            EXPECTED,
            left,
            replace(
                right,
                occurrences=right.occurrences
                + (replace(right.occurrences[0], position=10, clock=99),),
            ),
        ),
        compare(
            "changed-payload-old-attestation",
            TABLE,
            EXPECTED,
            left,
            update(right, "e2", receipt=changed),
        ),
        compare(
            "substituted-key",
            TABLE,
            EXPECTED,
            left,
            update(right, "e2", receipt=replace(RECEIPTS[2], Actor="C")),
        ),
        compare("authenticated-fork", fork_table, EXPECTED, left, forked),
        compare(
            "authenticated-causal-cycle",
            cycle_table,
            EXPECTED,
            update(left, "e0", receipt=cyclic, attestation="fixture-cycle"),
            update(right, "e0", receipt=cyclic, attestation="fixture-cycle"),
        ),
        compare("missing-parent", TABLE, EXPECTED, left, omit(right, "e0")),
        compare("omitted-expected-event", TABLE, EXPECTED, left, omit(right, "e3")),
        compare(
            "unavailable-verifier",
            TABLE,
            EXPECTED,
            left,
            update(right, "e2", attestation="unavailable"),
        ),
        compare(
            "conflict-with-missing-coverage",
            fork_table,
            EXPECTED,
            left,
            omit(forked, "e3"),
        ),
        compare(
            "different-observation-channel",
            TABLE | {(alternate, "fixture-e4")},
            EXPECTED + ("e4",),
            view("A", extended),
            view("B", extended),
        ),
        compare("coherent-fabricated-counterparties", TABLE, EXPECTED, left, right),
        compare(
            "coherent-fabricated-complete-graph",
            fabricated_table,
            tuple(r.EventId for r in fabricated),
            view("A", fabricated),
            view("B", fabricated),
        ),
        compare(
            "open-causal-boundary",
            boundary_table,
            EXPECTED,
            update(left, "e0", receipt=boundary, attestation="fixture-boundary"),
            update(right, "e0", receipt=boundary, attestation="fixture-boundary"),
        ),
        compare(
            "causal-coordinate-reversal",
            TABLE,
            EXPECTED,
            left,
            replace(
                right,
                occurrences=tuple(
                    replace(o, position=3 - o.position) for o in right.occurrences
                ),
            ),
        ),
        compare(
            "coordinate-collision",
            TABLE,
            EXPECTED,
            left,
            replace(
                right,
                occurrences=tuple(replace(o, position=0) for o in right.occurrences),
            ),
        ),
        compare("empty-cut", set(), (), view("A", ()), view("B", ())),
    ]


def entropy_panel() -> list[dict[str, Any]]:
    panels: list[tuple[str, list[tuple[int, ...]]]] = [
        ("independent", list(product(range(2), repeat=3))),
        ("copies", [(a, a, a) for a in range(2)]),
        ("inversion", [(a, 1 - a, a) for a in range(2)]),
        (
            "pairwise-independent-xor",
            [(a, b, a ^ b) for a, b in product(range(2), repeat=2)],
        ),
    ]
    output = []
    for name, rows in panels:
        n = len(rows)
        maximum = Fraction(max(Counter(rows).values()), n)
        pairwise = all(
            Fraction(sum(row[a] == x and row[b] == y for row in rows), n)
            == Fraction(sum(row[a] == x for row in rows), n)
            * Fraction(sum(row[b] == y for row in rows), n)
            for a, b in combinations(range(3), 2)
            for x, y in product(range(2), repeat=2)
        )
        premise = True
        for column in range(3):
            prefixes = {row[:column] for row in rows}
            for prefix in prefixes:
                group = [row for row in rows if row[:column] == prefix]
                premise &= max(
                    Counter(row[column] for row in group).values()
                ) * 2 <= len(group)
        denominator = maximum.denominator // maximum.numerator
        assert denominator > 0 and denominator & (denominator - 1) == 0
        output.append(
            {
                "Name": name,
                "Outcomes": [list(row) for row in rows],
                "MaximumProbability": [
                    str(maximum.numerator),
                    str(maximum.denominator),
                ],
                "JointMinEntropyBits": denominator.bit_length() - 1,
                "PairwiseIndependent": pairwise,
                "OneBitConditionalPremise": premise,
            }
        )
    return output


def scaling_panel() -> list[dict[str, Any]]:
    output = []
    for n, shape in product(
        (0, 1, 2, 3, 4, 8, 16, 32, 64), ("empty", "complete", "path", "star", "cycle")
    ):
        if shape == "complete":
            pairs = set(combinations(range(n), 2))
        elif shape == "path":
            pairs = {(i, i + 1) for i in range(n - 1)}
        elif shape == "star":
            pairs = {(0, i) for i in range(1, n)}
        elif shape == "cycle" and n >= 3:
            pairs = {(min(i, (i + 1) % n), max(i, (i + 1) % n)) for i in range(n)}
        else:
            pairs = set()
        baseline, relational = n * 32, len(pairs) * 8
        total = baseline + relational
        multiplier = Fraction(total, baseline) if baseline else None
        output.append(
            {
                "Name": f"{shape}-{n}",
                "Identities": n,
                "Required": len(pairs),
                "Observed": max(0, len(pairs) - 1),
                "Missing": int(bool(pairs)),
                "Baseline": str(baseline),
                "Relational": str(relational),
                "Total": str(total),
                "Multiplier": [str(multiplier.numerator), str(multiplier.denominator)]
                if multiplier
                else [],
            }
        )
    return output


def chsh(a: list[dict[str, int]], b: list[dict[str, int]]) -> float:
    buckets: dict[tuple[int, int], list[int]] = defaultdict(list)
    for ra, rb in zip(a, b, strict=True):
        buckets[ra["Setting"], rb["Setting"]].append(ra["Outcome"] * rb["Outcome"])
    return sum(
        ((-1 if key == (0, 1) else 1) * sum(values) / len(values))
        for key, values in buckets.items()
    )


def baseline_panel() -> list[dict[str, Any]]:
    a = [0, 0, 1, 1, 0, 0, 1, 1]
    code = [0, 1, 0, 1, 0, 1, 0, 1]
    chsh_a = [{"Setting": i // 2 % 2, "Outcome": 1} for i in range(8)]
    local = [{"Setting": i % 2, "Outcome": 1} for i in range(8)]
    conducted = [
        {"Setting": i % 2, "Outcome": -1 if i % 4 == 1 else 1} for i in range(8)
    ]
    cases = [
        ("exact-replay", True, a, local),
        ("inverted-replay", True, [1 - x for x in a], local),
        ("shared-seed-orthogonal-code", True, code, local),
        ("independent-finite-collision", False, a, local),
        ("distinct-controller-transcript", False, code, local),
        ("one-controller-same-transcript", True, code, local),
        ("conducted-chsh", True, code, conducted),
    ]
    output = []
    for name, shared, b, rounds in cases:
        correlation = abs(
            2 * sum(x == y for x, y in zip(a, b, strict=True)) / len(a) - 1
        )
        flag = correlation >= 0.8
        signature = (
            "accepted"
            if all(
                authenticate(TABLE, item) == "accepted"
                for item in view("A", RECEIPTS).occurrences
            )
            else "rejected"
        )
        output.append(
            {
                "Name": name,
                "SharedController": shared,
                "BitsA": a,
                "BitsB": b,
                "ChshA": chsh_a,
                "ChshB": rounds,
                "Correlation": correlation,
                "ReplayFlag": flag,
                "SignatureOnly": signature,
                "ReceiptStatus": compare(
                    "", TABLE, EXPECTED, view("A", RECEIPTS), view("B", RECEIPTS)
                )["Status"],
                "Spectrum": [max(abs(chsh(chsh_a, rounds)), abs(chsh(rounds, chsh_a)))],
                "ReplayFalsePositive": flag and not shared,
                "ReplayFalseNegative": shared and not flag,
            }
        )
    return output


def semantic_panel() -> dict[str, Any]:
    return {
        "Transport": transport_panel(),
        "Mutations": mutation_panel(),
        "Entropy": entropy_panel(),
        "Scaling": scaling_panel(),
        "Baselines": baseline_panel(),
    }


def verify_saved(
    saved: dict[str, Any], root: Path, *, check_source_snapshot: bool = True
) -> dict[str, Any]:
    if (
        saved.get("Protocol") != "relational-identity-v1"
        or saved.get("SourceArchive")
        != "archive/relational-identity-20260906-source-v1"
    ):
        raise ValueError("unexpected protocol or source archive")
    if saved.get("ProtocolCommit") != "4f470f40e":
        raise ValueError("unexpected preregistration commit")
    expected_paths = {
        "docs/research/relational-identity/2026-09-06-protocol.md",
        "docs/research/relational-identity/2026-09-06-clarification.md",
        "src/Research.FSharp/RelationalIdentity.fs",
        "src/Research.FSharp/RelationalIdentityExperiment.fs",
        "src/Research.FSharp/run-relational-identity.fsx",
        "src/Interp.Python/zeta_interp/relational_identity.py",
        "src/Core/AntiSybil.fs",
        "src/Core/CoordinationSpectrum.fs",
    }
    hashes = saved.get("SourceHashes", [])
    if (
        len(hashes) != len(expected_paths)
        or {row["Path"] for row in hashes} != expected_paths
    ):
        raise ValueError("source hash coverage differs from registered files")
    for row in hashes:
        if (
            sha256((root / row["Path"]).read_bytes()).hexdigest().upper()
            != row["Sha256"]
        ):
            raise ValueError("source hash mismatch: " + row["Path"])
    expected = semantic_panel()
    if saved.get("Semantic") != expected:
        raise ValueError(
            "native receipt differs from independently regenerated semantic panel"
        )
    return {
        "Protocol": "relational-identity-independent-replay-v1",
        "SourceArchive": saved["SourceArchive"],
        "ExactSemanticMatch": True,
        "SourceSnapshotVerified": check_source_snapshot,
        "TransportChecks": expected["Transport"]["Checks"],
        "MutationCases": len(expected["Mutations"]),
        "EntropyCases": len(expected["Entropy"]),
        "ScalingCases": len(expected["Scaling"]),
        "BaselineCases": len(expected["Baselines"]),
    }


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("receipt", type=Path)
    parser.add_argument("output", type=Path)
    args = parser.parse_args()
    if args.output.exists():
        parser.error("refusing to overwrite replay output")
    root = Path(__file__).resolve().parents[3]
    result = verify_saved(json.loads(args.receipt.read_text()), root)
    result["NativeReceiptSha256"] = (
        sha256(args.receipt.read_bytes()).hexdigest().upper()
    )
    args.output.write_text(json.dumps(result, indent=2) + "\n")


if __name__ == "__main__":
    main()

File: src/Research.FSharp/RelationalIdentity.fs
namespace Zeta.Research

open System

/// Finite research receipts. Authentication and admission policy belong to the caller.
[<RequireQualifiedAccess>]
module RelationalIdentity =
    type Receipt =
        { EventId: string
          Actor: string
          Counterparty: string
          Interaction: string
          Channel: string
          Claim: string
          Weight: int64
          Parents: string list }

    type Occurrence = { Position: int64; Clock: int64; Receipt: Receipt; Attestation: string }
    type View = { Observer: string; Occurrences: Occurrence list }
    type Authentication = Accepted | Rejected | Unavailable
    type Refusal = InvalidInput of string | InvalidCoordinates of observer: string
    type ClaimBalance = { Actor: string; Counterparty: string; Channel: string; Claim: string; Weight: bigint }
    type Invariant = { Receipts: Receipt list; CausalPairs: (string * string) list; Claims: ClaimBalance list }
    type Readout =
        { Status: string
          Expected: string list
          MissingLeft: string list
          MissingRight: string list
          UnverifiedLeft: string list
          UnverifiedRight: string list
          RejectedLeft: string list
          RejectedRight: string list
          Conflicts: string list
          BoundaryParents: string list
          RepeatedLeft: int
          RepeatedRight: int
          Invariant: Invariant option }

    type private CheckedView =
        { View: View
          Accepted: Occurrence list
          Present: Set<string>
          Unverified: Set<string>
          Rejected: Set<string>
          Repeated: int }

    let private validId (value: string) = not (String.IsNullOrWhiteSpace value) && value.Length <= 128
    let private present (value: 'a) = not (isNull (box value))

    /// Stable parent-set encoding. Signature capabilities receive this exact typed content.
    let canonical (receipt: Receipt) = { receipt with Parents = List.distinct receipt.Parents |> List.sort }

    let private validate cut (view: View) =
        if not (validId view.Observer) || not (present view.Occurrences) || view.Occurrences.Length > 256 then
            Error(InvalidInput "view requires an observer and at most 256 occurrences")
        elif view.Occurrences |> List.exists (fun item ->
            not (present item) || not (present item.Receipt)
            || not (present item.Receipt.Parents) || item.Receipt.Parents.Length > 128
            || [ item.Receipt.EventId; item.Receipt.Actor; item.Receipt.Counterparty
                 item.Receipt.Interaction; item.Receipt.Channel; item.Receipt.Claim ] |> List.exists (validId >> not)
            || item.Receipt.Parents |> List.exists (validId >> not)
            || not (present item.Attestation) || item.Attestation.Length > 128) then
            Error(InvalidInput "receipt requires bounded typed IDs, parents and attestation")
        elif view.Occurrences |> List.exists (fun item -> not (Set.contains item.Receipt.EventId cut)) then
            Error(InvalidInput "occurrences must belong to the declared cut")
        elif view.Occurrences |> List.exists (fun item -> item.Position < 0L)
             || (view.Occurrences |> List.map _.Position |> List.distinct |> List.length) <> view.Occurrences.Length then
            Error(InvalidCoordinates view.Observer)
        else Ok view

    let private authenticate verify (view: View) =
        let items = view.Occurrences |> List.map (fun item -> { item with Receipt = canonical item.Receipt })
        let classified = items |> List.map (fun item -> item, verify item.Receipt item.Attestation)
        let ids state = classified |> List.choose (fun (item, outcome) -> if outcome = state then Some item.Receipt.EventId else None) |> Set.ofList
        let accepted = classified |> List.choose (fun (item, outcome) -> if outcome = Accepted then Some item else None)
        { View = view; Accepted = accepted; Present = items |> List.map _.Receipt.EventId |> Set.ofList
          Unverified = ids Unavailable; Rejected = ids Rejected
          Repeated = accepted.Length - (accepted |> List.map _.Receipt |> List.distinct |> List.length) }

    let private closure (receipts: Receipt list) =
        let events = receipts |> List.map _.EventId |> Set.ofList
        let initial = receipts |> List.collect (fun receipt ->
            receipt.Parents |> List.choose (fun parent -> if Set.contains parent events then Some(parent, receipt.EventId) else None)) |> Set.ofList
        let mutable pairs = initial
        for pivot in events do
            let before = pairs |> Set.filter (fun (_, target) -> target = pivot) |> Set.map fst
            let after = pairs |> Set.filter (fun (source, _) -> source = pivot) |> Set.map snd
            pairs <- Set.union pairs (seq { for source in before do for target in after do yield source, target } |> Set.ofSeq)
        pairs

    let private balances (receipts: Receipt list) =
        receipts
        |> List.groupBy (fun receipt -> receipt.Actor, receipt.Counterparty, receipt.Channel, receipt.Claim)
        |> List.choose (fun ((actor, counterparty, channel, claim), group) ->
            let weight = group |> List.sumBy (fun receipt -> bigint receipt.Weight)
            if weight.IsZero then None
            else Some { Actor = actor; Counterparty = counterparty; Channel = channel; Claim = claim; Weight = weight })
        |> List.sortBy (fun item -> item.Actor, item.Counterparty, item.Channel, item.Claim)

    let private ordered (pairs: Set<string * string>) (view: CheckedView) =
        // Replays are observations of an old event, not a new causal position for that event.
        let positions = view.Accepted |> List.groupBy (fun item -> item.Receipt.EventId)
                        |> List.map (fun (event, items) -> event, items |> List.map _.Position |> List.min) |> Map.ofList
        pairs |> Set.forall (fun (source, target) ->
            match Map.tryFind source positions, Map.tryFind target positions with
            | Some before, Some after -> before < after
            | _ -> true)

    /// Compare only the caller-declared cut. Non-detection never asserts distinctness.
    /// The supplied verifier must be total; this kernel implements no cryptographic primitive.
    let compareViews verify (expected: string list) (left: View) (right: View) =
        if not (present expected) || expected.Length > 128 || expected |> List.exists (validId >> not)
           || not (present left) || not (present right) then
            Error(InvalidInput "a cut requires at most 128 valid event IDs and two views")
        else
            let cut = Set.ofList expected
            match validate cut left, validate cut right with
            | Error error, _ | _, Error error -> Error error
            | Ok _, Ok _ ->
                let a, b = authenticate verify left, authenticate verify right
                let groups = (a.Accepted @ b.Accepted) |> List.map _.Receipt |> List.groupBy _.EventId
                let conflicts = groups |> List.choose (fun (event, items) -> if (List.distinct items).Length > 1 then Some event else None) |> List.sort
                let receipts = groups |> List.collect snd |> List.distinct |> List.sort
                let pairs = closure receipts
                let cycle = pairs |> Set.exists (fun (source, target) -> source = target)
                let boundary = receipts |> List.collect _.Parents |> Set.ofList |> fun parents -> Set.difference parents cut |> Set.toList
                let missing view = Set.difference cut view.Present |> Set.toList
                let unknown = not (Set.isSubset cut (a.Accepted |> List.map _.Receipt.EventId |> Set.ofList))
                              || not (Set.isSubset cut (b.Accepted |> List.map _.Receipt.EventId |> Set.ofList))
                              || not (List.isEmpty boundary)
                              || not (Set.isEmpty a.Unverified && Set.isEmpty b.Unverified)
                if List.isEmpty conflicts && not cycle && (not (ordered pairs a) || not (ordered pairs b)) then
                    Error(InvalidCoordinates(if not (ordered pairs a) then left.Observer else right.Observer))
                else
                    let status =
                        if not (List.isEmpty conflicts) then "authenticated-conflict"
                        elif cycle then "authenticated-causal-cycle"
                        elif not (Set.isEmpty a.Rejected && Set.isEmpty b.Rejected) then "authentication-rejected"
                        elif unknown then "unknown-coverage"
                        else "consistent-on-declared-cut"
                    Ok { Status = status; Expected = Set.toList cut
                         MissingLeft = missing a; MissingRight = missing b
                         UnverifiedLeft = Set.toList a.Unverified; UnverifiedRight = Set.toList b.Unverified
                         RejectedLeft = Set.toList a.Rejected; RejectedRight = Set.toList b.Rejected
                         Conflicts = conflicts; BoundaryParents = boundary
                         RepeatedLeft = a.Repeated; RepeatedRight = b.Repeated
                         Invariant = if status = "consistent-on-declared-cut" then Some { Receipts = receipts; CausalPairs = Set.toList pairs; Claims = balances receipts } else None }

    /// Coordinate transport is a candidate until compareViews validates the resulting chart.
    let rechart positions clocks (view: View) =
        { view with Occurrences = view.Occurrences |> List.map (fun item -> { item with Position = positions item.Position; Clock = clocks item.Clock }) }

    type Workload =
        { Identities: int
          Required: int
          Observed: int
          Missing: int
          Baseline: bigint
          Relational: bigint
          Total: bigint
          Multiplier: (bigint * bigint) option }

    /// Exact stipulated accounting, not a measured lower bound on an adaptive adversary.
    let workload (identityCosts: (string * bigint) list) (required: (string * string) list) (observed: (string * string) list) pairCost =
        if not (present identityCosts) || not (present required) || not (present observed)
           || identityCosts.Length > 64 || required.Length > 4096 || observed.Length > 4096
           || pairCost < 0I || identityCosts |> List.exists (fun (identity, cost) -> not (validId identity) || cost <= 0I)
           || (identityCosts |> List.map fst |> List.distinct |> List.length) <> identityCosts.Length then
            Error(InvalidInput "workload requires distinct identities, positive source costs and nonnegative pair cost")
        else
            let identities = identityCosts |> List.map fst |> Set.ofList
            let invalid (a, b) = a = b || not (Set.contains a identities) || not (Set.contains b identities)
            if List.exists invalid required || List.exists invalid observed then Error(InvalidInput "pair endpoints must be distinct declared identities")
            else
                let edge (a, b) = if String.CompareOrdinal(a, b) < 0 then a, b else b, a
                let needed = required |> List.map edge |> Set.ofList
                let seen = observed |> List.map edge |> Set.ofList
                if not (Set.isSubset seen needed) then Error(InvalidInput "observed obligations must belong to the declared required graph")
                else
                    let baseline = identityCosts |> List.sumBy snd
                    let relational = bigint needed.Count * pairCost
                    let total = baseline + relational
                    let multiplier =
                        if baseline.IsZero then None
                        else
                            let divisor = System.Numerics.BigInteger.GreatestCommonDivisor(total, baseline)
                            Some(total / divisor, baseline / divisor)
                    Ok { Identities = identities.Count; Required = needed.Count; Observed = seen.Count
                         Missing = Set.difference needed seen |> Set.count; Baseline = baseline
                         Relational = relational; Total = total; Multiplier = multiplier }
